tjpcov/nmt_tools.py

```diff
@@ -602,37 +602,44 @@ def get_nell(sacc_data, bins=None, nside=None, cache=None):
 
     Returns:
     --------
-        nell (int): Number of ells for the fidicual Cells points; i.e. 3*nside
-    """
-    try:
-        dtype = sacc_data.get_data_types()[0]
-        tracers = sacc_data.get_tracer_combinations(data_type=dtype)[0]
-        ix = sacc_data.indices(data_type=dtype, tracers=tracers)
-        bpw = sacc_data.get_bandpower_windows(ix)
-        if bpw is None:
-            raise ValueError
-        nell = bpw.nell
-    except ValueError as e:
-        # If the window functions are wrong. Do magic
-        warnings.warn('The window functions in the sacc file are wrong: ')
-        warnings.warn(str(e))
-        warnings.warn('Circunventing this error')
-
-        if bins is not None:
-            nell = bins.lmax + 1
-        elif nside is not None:
-            nell = 3*nside
-        elif 'workspaces' in cache:
-            w = list(list(cache['workspaces'].values())[0].values())[0]
-            if isinstance(w, nmt.NmtWorkspace):
-                bpw = w.get_bandpower_windows()
-                nell = bpw.shape[-1]
+        nell (int): Number of ells for the fidicual Cells points; i.e. lmax or
+        3*nside
+    """
+    # Extracting the workspace from the cache first to use it later in a easy
+    # way.
+    if (cache is not None) and ('workspaces' in cache):
+        w = list(list(cache['workspaces'].values())[0].values())[0]
+    else:
+        w = None
+
+    if isinstance(w, nmt.NmtWorkspace):
+        # We don't want to read the workspace just to get the nell
+        bpw = w.get_bandpower_windows()
+        nell = bpw.shape[-1]
+    elif bins is not None:
+        nell = bins.lmax + 1
+    else:
+        try:
+            dtype = sacc_data.get_data_types()[0]
+            tracers = sacc_data.get_tracer_combinations(data_type=dtype)[0]
+            ix = sacc_data.indices(data_type=dtype, tracers=tracers)
+            bpw = sacc_data.get_bandpower_windows(ix)
+            if bpw is None:
+                raise ValueError
+            nell = bpw.nell
+        except ValueError as e:
+            # If the window functions are wrong. Do magic
+            warnings.warn('The window functions in the sacc file are wrong: ')
+            warnings.warn(str(e))
+            if "binning/ell_max" in sacc_data.metadata:
+                nell = sacc_data.metadata["binning/ell_max"] + 1
+            elif nside is not None:
```

Comment:
Would putting the check for `nside` first make more sense here?

Reply:
I started implementing `nside` first, but then i realized that I'm not sure it will work or fail if `lmax < 3*nside`.

Reply:
Could the situation happen that `binning/ell_max > 3*nside`? My feeling is that unless someone really knows what they're doing, `lmax = 3*nside` is the safe default.
In any case, having the warning about which `nell` is being used (currently on L637) for both of the `if`-statement would be useful.

Reply:
I think having `nell > 3*nside` should be fine. If I remember correctly, NaMaster does not take into account those ells, anyway. In any case, if have added a new check so that if `nside` is given and `binning/lmax > 3*nside-1`, one uses `nell = 3*nside`. 

I have also added the warning you mentioned.